Add an example (#35)
unicorn-wg/oblivious-http: draft-thomson-http-oblivious.md

--- back

# Complete Example of a Request and Response

<!-- Generated using ohttp:
RUST_LOG=ohttp cargo test -p ohttp -\-lib -\- -\-nocapture request_response
-->

A single request and response exchange is shown here. Binary values (key
configuration, secret keys, the content of messages, and intermediate values)
are shown in hexadecimal. The request and response here are absolutely minimal;
the purpose of this example is to show the cryptographic operations.

The oblivious request resource generates a key pair. In this example the server

> chooses the DHKEM(X25519, HKDF-SHA256) and generates an X25519 key pair

Reviewer: ```suggestion
chooses DHKEM(X25519, HKDF-SHA256) and generates an X25519 key pair
```

{{?X25519=RFC7748}}. The X25519 secret key is:

~~~
15ce887006e079dcc7d67e73e5c13e31a55083f816eca9ebcf523b90ea2ab7b0
~~~

The oblivious request resource constructs a key configuration that includes the
corresponding public key as follows:

~~~
0100200020f21c612398e4384c21b7f2a759133c6c2d1b9ce6d033613dfad2c7
3d4826214900080001000100010003
~~~

This key configuration is somehow obtained by the client. Then when a client
wishes to send an HTTP request of a GET request to `https://example.com`, it constructs the following binary HTTP message:

~~~
00034745540568747470730b6578616d706c652e636f6d012f
~~~

> The client then reads the oblivious request resource key configuration, selects a pair of KDF and AEAD and generates an X25519 key pair:

Reviewer: ```suggestion
The client then reads the oblivious request resource key configuration and selects a mutually supported KDF and AEAD. In this example, the client selects HKDF-SHA256 and AES-128-GCM. The client then generates an X25519 key pair:
```

Reviewer: I'm not really sure how I feel about this particular part, since it exposes HPKE internals. Maybe we could rewrite this to something like so:

```
The client then generates an HPKE context using the server's public key, yielding an encapsulated 
key:

<X25519 public key>

With the corresponding private key:

<X25519 private key>
```

Author: Agree. Will fix.

~~~ http-message
POST /request.example.net/proxy HTTP/1.1
Host: proxy.example.org
Content-Type: message/ohttp-req
Content-Length: 78

<content is the encapsulated request above>
~~~

The oblivious proxy resource receives this request and forwards it to the
oblivious request resource, which might look like:

~~~ http-message
POST /oblivious/request HTTP/1.1
Host: example.com
Content-Type: message/ohttp-req
Content-Length: 78

<content is the encapsulated request above>
~~~

The oblivous request resource receives this request, selects the key it
generated previously using the key identifier from the message, and decrypts
the message. As this request is directed to the same server, the oblivious
request resource does not need to initiate an HTTP request to the oblivious
target resource. The request can be served directly by the oblivious target
resource, which generates a minimal response (consisting of just a 200 status
code) as follows:

~~~
0140c8
~~~

The response is constructed by extracting a secret from the HPKE context:

~~~
861c67eefd91906068ee3208a9102274
~~~

The key derivation for the encapsulated response uses both the encapsulated KEM

> key from the request and a randomly selected value. This produces a salt of:

Reviewer: ```suggestion
key from the request and a randomly selected nonce. This produces a salt of:
```

Reviewer: (for consistency with the terminology used in the main text)

~~~
a7e0dffe93bc9ed807b51588f10669cd2f09ceb7f6a71153658275993eb88276
55645b6626d6e8b1163d739c6cd3d94a
~~~

The salt and secret are both passed to the Extract function of the selected KDF
(HKDF-SHA256) to produce a pseudorandom key of:

~~~
ac29e0e1df48de4b349bb97db74da7f6732557a2fa6d12019ddadd8bf1e9cd99
~~~

The pseudorandom key is used with the Expand function of the KDF and an info
field of "key" to produce a 16-byte key for the selected AEAD (AES-128-GCM):

~~~
0d81ce961eb6e61d510e18053418f316
~~~

With the same KDF and pseudorandom key, an info field of "nonce" is used to
generate a 12-byte nonce:

~~~
ba2d3637b5f041cc15ef73b2
~~~

> The AEAD Seal function is then used to encrypted the response, which is added

Reviewer: ```suggestion
The AEAD Seal function is then used to encrypt the response, which is added
```

to the randomized nonce value to produce the encapsulated response:

~~~
55645b6626d6e8b1163d739c6cd3d94ac7e0cfd48adf8057517241ee220d20bc
e00a4a
~~~

The oblivious request resource then constructs a response:

~~~ http-message
HTTP/1.1 200 OK
Date: Wed, 27 Jan 2021 04:45:07 GMT
Cache-Control: private, no-store
Content-Type: message/ohttp-res
Content-Length: 38

<content is the encapsulated response>
~~~

The same response might then be generated by the oblivious proxy resource which
might change as little as the Date header. The client is then able to use the
HPKE context it created and the nonce from the encapsulated response to

> construct the AEAD key and nonce and extract the response.

Reviewer: ```suggestion
construct the AEAD key and nonce and decrypt the response.
```
